File: python/quantize.py
"""
Bit-Level Weight Quantization Engine — implemented from scratch.

What this demonstrates
----------------------
Quantization compresses model weights from float32/float16 into fewer bits
(8-bit, 4-bit, 2-bit) to reduce memory and speed up inference. Libraries like
bitsandbytes hide the math. This file performs the actual bit math in pure
NumPy so you can see exactly where precision is lost.

Schemes implemented
-------------------
1. Absmax (symmetric)   — used in LLM.int8() for 8-bit
2. Zero-point (affine)  — used in GPTQ-style 4-bit
3. Block quantization   — divide matrix into blocks; each block has its own
                          scale (reduces outlier sensitivity)
4. NF4 (Normal Float 4) — used in QLoRA; non-uniform grid matched to
                          normal distribution of weights

Run
---
    python quantize.py

Dependencies: numpy only.
"""

from __future__ import annotations

import struct
import numpy as np

RNG = np.random.default_rng(0)

# ──────────────────────────────────────────────────────────────────────────────
# Helper: float16 bit representation
# ──────────────────────────────────────────────────────────────────────────────

def float_to_bits(x: float, dtype: str = "f2") -> str:
    """Return binary string for a float16 (e) or float32 (f) value."""
    fmt    = "e" if dtype == "f2" else "f"
    size   = 2   if dtype == "f2" else 4
    packed = struct.pack(f">{fmt}", x)
    return " ".join(f"{b:08b}" for b in packed[:size])


# ──────────────────────────────────────────────────────────────────────────────
# 1. Absmax symmetric quantization  (8-bit or any n-bit)
# ──────────────────────────────────────────────────────────────────────────────

def absmax_quantize(
    weights: np.ndarray, bits: int = 8
) -> tuple[np.ndarray, float, np.ndarray]:
    """
    Symmetric quantization: maps [-max_abs, +max_abs] → [-2^(bits-1), 2^(bits-1)-1].

    scale  = max(|W|) / (2^(bits-1) - 1)
    W_int  = round(W / scale)
    W_deq  = W_int * scale
    """
    levels   = 2 ** (bits - 1) - 1           # e.g. 127 for 8-bit
    max_abs  = float(np.abs(weights).max())
    scale    = max_abs / levels if max_abs != 0 else 1.0

    quantized   = np.round(weights / scale).clip(-levels, levels).astype(np.int8)
    dequantized = quantized.astype(np.float32) * scale

    return quantized, scale, dequantized


# ──────────────────────────────────────────────────────────────────────────────
# 2. Zero-point affine quantization (asymmetric)
# ──────────────────────────────────────────────────────────────────────────────

def zeropoint_quantize(
    weights: np.ndarray, bits: int = 4
) -> tuple[np.ndarray, float, int, np.ndarray]:
    """
    Asymmetric quantization: maps [min(W), max(W)] → [0, 2^bits - 1].

    scale      = (max - min) / (2^bits - 1)
    zero_point = round(-min / scale)
    W_int      = round(W / scale) + zero_point
    W_deq      = (W_int - zero_point) * scale
    """
    levels     = 2 ** bits - 1                # e.g. 15 for 4-bit
    w_min, w_max = float(weights.min()), float(weights.max())
    scale      = (w_max - w_min) / levels if (w_max - w_min) != 0 else 1.0
    zero_point = int(round(-w_min / scale))

    quantized   = (np.round(weights / scale) + zero_point).clip(0, levels).astype(np.uint8)
    dequantized = (quantized.astype(np.float32) - zero_point) * scale

    return quantized, scale, zero_point, dequantized


# ──────────────────────────────────────────────────────────────────────────────
# 3. Block quantization
# ──────────────────────────────────────────────────────────────────────────────

def block_quantize(
    weights: np.ndarray, bits: int = 4, block_size: int = 4
) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """
    Divide weight vector into blocks; each block gets its own scale.
    Reduces sensitivity to large outlier weights.

    Returns
    -------
    quantized   — integer representation (same shape)
    scales      — one scale per block
    dequantized — reconstructed float weights
    """
    flat     = weights.flatten()
    n        = len(flat)
    levels   = 2 ** (bits - 1) - 1
    n_blocks = (n + block_size - 1) // block_size

    quantized   = np.zeros(n, dtype=np.int8)
    dequantized = np.zeros(n, dtype=np.float32)
    scales      = np.zeros(n_blocks, dtype=np.float32)

    for b in range(n_blocks):
        start, end = b * block_size, min((b + 1) * block_size, n)
        block      = flat[start:end].astype(np.float32)
        max_abs    = float(np.abs(block).max())
        scale      = max_abs / levels if max_abs != 0 else 1.0
        scales[b]  = scale

        q = np.round(block / scale).clip(-levels, levels).astype(np.int8)
        quantized[start:end]   = q
        dequantized[start:end] = q.astype(np.float32) * scale

    return quantized.reshape(weights.shape), scales, dequantized.reshape(weights.shape)


# ──────────────────────────────────────────────────────────────────────────────
# 4. NF4 — Normal Float 4 (QLoRA)
# ──────────────────────────────────────────────────────────────────────────────

# 16 quantile levels derived from the standard normal CDF so the grid is
# denser where normal-distributed weights cluster (near zero).
_NF4_LEVELS = np.array([
    -1.0,       -0.6961,    -0.5251,    -0.3951,
    -0.2844,    -0.1848,    -0.0917,     0.0,
     0.0796,     0.1609,     0.2461,     0.3379,
     0.4407,     0.5626,     0.7230,     1.0,
], dtype=np.float32)


def nf4_quantize(
    weights: np.ndarray, absmax: float | None = None
) -> tuple[np.ndarray, float, np.ndarray]:
    """
    NF4 as used in QLoRA (Dettmers et al., 2023).

    Normalise weights to [-1, 1] then map each value to the nearest level in
    the 16-point NF4 grid. Store as 4-bit indices (uint8 here for clarity).
    """
    if absmax is None:
        absmax = float(np.abs(weights).max()) or 1.0

    normalised  = weights.astype(np.float32) / absmax
    # Vectorised nearest-neighbour in the NF4 grid
    indices     = np.abs(normalised[..., None] - _NF4_LEVELS).argmin(axis=-1).astype(np.uint8)
    dequantized = _NF4_LEVELS[indices] * absmax

    return indices, absmax, dequantized


# ──────────────────────────────────────────────────────────────────────────────
# Metrics
# ──────────────────────────────────────────────────────────────────────────────

def mse(original: np.ndarray, reconstructed: np.ndarray) -> float:
    return float(np.mean((original.astype(np.float32) - reconstructed.astype(np.float32)) ** 2))


def snr_db(original: np.ndarray, reconstructed: np.ndarray) -> float:
    """Signal-to-noise ratio in dB (higher = better reconstruction)."""
    signal_power = float(np.mean(original.astype(np.float32) ** 2))
    noise_power  = mse(original, reconstructed)
    if noise_power == 0:
        return float("inf")
    return float(10 * np.log10(signal_power / noise_power))


def memory_bytes(bits: int, n_elements: int) -> float:
    return (bits * n_elements) / 8


# ──────────────────────────────────────────────────────────────────────────────
# Bit inspector
# ──────────────────────────────────────────────────────────────────────────────

def show_bit_comparison(original: float, quantized_int: int, reconstructed: float, bits: int) -> None:
    orig_bits  = float_to_bits(float(np.float16(original)), dtype="f2")
    recon_bits = float_to_bits(float(np.float16(reconstructed)), dtype="f2")
    q_bits     = f"{int(quantized_int) & ((1 << bits) - 1):0{bits}b}"
    print(f"    original  {original:+.6f}  → float16: {orig_bits}")
    print(f"    quantized                  → int{bits}:    {q_bits}")
    print(f"    recon'd   {reconstructed:+.6f}  → float16: {recon_bits}")
    print(f"    bit error: |{original:.6f} - {reconstructed:.6f}| = {abs(original - reconstructed):.6f}")


# ──────────────────────────────────────────────────────────────────────────────
# Main demo
# ──────────────────────────────────────────────────────────────────────────────

def main() -> None:
    HR = "=" * 65
    print(HR)
    print("  Bit-Level Weight Quantization Engine — from scratch (NumPy)")
    print(HR)

    # Simulate a small weight matrix (e.g., one FFN layer row)
    weights = RNG.normal(0, 0.02, (4, 8)).astype(np.float32)
    n_el    = weights.size
    fp32_bytes = memory_bytes(32, n_el)

    print(f"\nWeight matrix shape: {weights.shape}  ({n_el} elements)")
    print(f"Original (float32) memory: {fp32_bytes:.0f} bytes")
    print(f"\nSample weights:\n{weights.round(5)}")

    # ── 8-bit absmax ──────────────────────────────────────────────────────────
    print(f"\n{'─'*65}")
    print("  [1] Absmax Symmetric — 8-bit")
    print(f"{'─'*65}")
    q8, scale8, deq8 = absmax_quantize(weights, bits=8)
    print(f"  Scale factor : {scale8:.8f}")
    print(f"  Quantized    :\n{q8}")
    print(f"  Dequantized  :\n{deq8.round(5)}")
    print(f"  MSE          : {mse(weights, deq8):.2e}")
    print(f"  SNR          : {snr_db(weights, deq8):.1f} dB")
    print(f"  Memory       : {memory_bytes(8, n_el):.0f} bytes  ({fp32_bytes/memory_bytes(8,n_el):.1f}x smaller)")
    print("\n  Bit-level inspection (first weight):")
    show_bit_comparison(float(weights.flat[0]), int(q8.flat[0]), float(deq8.flat[0]), bits=8)

    # ── 4-bit zero-point ──────────────────────────────────────────────────────
    print(f"\n{'─'*65}")
    print("  [2] Zero-Point Affine — 4-bit")
    print(f"{'─'*65}")
    q4, scale4, zp4, deq4 = zeropoint_quantize(weights, bits=4)
    print(f"  Scale        : {scale4:.8f}   Zero-point: {zp4}")
    print(f"  Quantized    :\n{q4}")
    print(f"  Dequantized  :\n{deq4.round(5)}")
    print(f"  MSE          : {mse(weights, deq4):.2e}")
    print(f"  SNR          : {snr_db(weights, deq4):.1f} dB")
    print(f"  Memory       : {memory_bytes(4, n_el):.0f} bytes  ({fp32_bytes/memory_bytes(4,n_el):.1f}x smaller)")
    print("\n  Bit-level inspection (first weight):")
    show_bit_comparison(float(weights.flat[0]), int(q4.flat[0]), float(deq4.flat[0]), bits=4)

    # ── 4-bit block quantization ──────────────────────────────────────────────
    print(f"\n{'─'*65}")
    print("  [3] Block Quantization — 4-bit, block_size=4")
    print(f"{'─'*65}")
    qb, scales_b, deqb = block_quantize(weights, bits=4, block_size=4)
    print(f"  Per-block scales: {scales_b.round(8)}")
    print(f"  Quantized    :\n{qb}")
    print(f"  Dequantized  :\n{deqb.round(5)}")
    print(f"  MSE          : {mse(weights, deqb):.2e}")
    print(f"  SNR          : {snr_db(weights, deqb):.1f} dB")
    overhead = scales_b.nbytes
    print(f"  Memory       : {memory_bytes(4, n_el) + overhead:.0f} bytes (incl. scale overhead)")

    # ── NF4 ───────────────────────────────────────────────────────────────────
    print(f"\n{'─'*65}")
    print("  [4] NF4 — Normal Float 4 (QLoRA)")
    print(f"{'─'*65}")
    qnf4, absmax_nf4, deq_nf4 = nf4_quantize(weights)
    print(f"  Absmax       : {absmax_nf4:.8f}")
    print(f"  NF4 grid (16 levels): {_NF4_LEVELS.round(4).tolist()}")
    print(f"  Indices      :\n{qnf4}")
    print(f"  Dequantized  :\n{deq_nf4.round(5)}")
    print(f"  MSE          : {mse(weights, deq_nf4):.2e}")
    print(f"  SNR          : {snr_db(weights, deq_nf4):.1f} dB")
    print(f"  Memory       : {memory_bytes(4, n_el):.0f} bytes  ({fp32_bytes/memory_bytes(4,n_el):.1f}x smaller)")

    # ── Summary table ─────────────────────────────────────────────────────────
    print(f"\n{HR}")
    print("  Summary: precision vs. compression")
    print(HR)
    print(f"  {'Method':<30} {'Bits':>4}  {'MSE':>10}  {'SNR (dB)':>9}  {'Memory':>7}")
    print(f"  {'──────':<30} {'────':>4}  {'───':>10}  {'────────':>9}  {'──────':>7}")
    rows = [
        ("float32 (original)",   32, 0.0,                    float("inf"),          fp32_bytes),
        ("Absmax 8-bit",          8, mse(weights, deq8),      snr_db(weights, deq8),  memory_bytes(8, n_el)),
        ("ZeroPoint 4-bit",       4, mse(weights, deq4),      snr_db(weights, deq4),  memory_bytes(4, n_el)),
        ("Block 4-bit (b=4)",     4, mse(weights, deqb),      snr_db(weights, deqb),  memory_bytes(4, n_el)),
        ("NF4 4-bit (QLoRA)",     4, mse(weights, deq_nf4),   snr_db(weights, deq_nf4), memory_bytes(4, n_el)),
    ]
    for name, bits, m, s, mem in rows:
        snr_str = f"{s:.1f}" if s != float("inf") else "∞"
        print(f"  {name:<30} {bits:>4}  {m:>10.2e}  {snr_str:>9}  {mem:>5.0f} B")
    print(HR)


if __name__ == "__main__":
    main()

File: typescript/quantize.ts
/**
 * Bit-Level Weight Quantization Engine — implemented from scratch.
 *
 * Four quantization schemes, pure TypeScript, zero dependencies:
 *   1. Absmax symmetric  (8-bit)
 *   2. Zero-point affine (4-bit)
 *   3. Block quantization (4-bit)
 *   4. NF4 — Normal Float 4 (QLoRA)
 *
 * Run:  npx ts-node quantize.ts  |  tsc && node quantize.js  |  deno run quantize.ts
 */

// ──────────────────────────────────────────────────────────────────────────────
// Seeded LCG (reproducible without library)
// ──────────────────────────────────────────────────────────────────────────────
class LCG {
  private state: number;
  constructor(seed = 0) { this.state = seed; }

  next(): number {
    this.state = (1664525 * this.state + 1013904223) & 0xffffffff;
    return (this.state >>> 0) / 0x100000000;
  }

  /** Normal-distributed sample via Box-Muller */
  normal(mean = 0, std = 1): number {
    const u1 = this.next() || 1e-10;
    const u2 = this.next();
    const z  = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
    return mean + std * z;
  }

  matrix(rows: number, cols: number, std = 0.02): number[][] {
    return Array.from({ length: rows }, () =>
      Array.from({ length: cols }, () => this.normal(0, std))
    );
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Flat view helpers
// ──────────────────────────────────────────────────────────────────────────────
function flatten(m: number[][]): number[] {
  return m.flat();
}

function reshape(flat: number[], rows: number, cols: number): number[][] {
  return Array.from({ length: rows }, (_, r) => flat.slice(r * cols, (r + 1) * cols));
}

function maxAbs(arr: number[]): number {
  return arr.reduce((m, x) => Math.max(m, Math.abs(x)), 0);
}

function minOf(arr: number[]): number { return Math.min(...arr); }
function maxOf(arr: number[]): number { return Math.max(...arr); }

function mse(a: number[], b: number[]): number {
  return a.reduce((s, v, i) => s + (v - b[i]) ** 2, 0) / a.length;
}

function snrDb(original: number[], reconstructed: number[]): number {
  const sp = original.reduce((s, v) => s + v * v, 0) / original.length;
  const np_ = mse(original, reconstructed);
  return np_ === 0 ? Infinity : 10 * Math.log10(sp / np_);
}

function memBytes(bits: number, n: number): number { return (bits * n) / 8; }

// ──────────────────────────────────────────────────────────────────────────────
// 1. Absmax symmetric quantization
// ──────────────────────────────────────────────────────────────────────────────
interface AbsmaxResult {
  quantized:   Int8Array;
  scale:       number;
  dequantized: number[];
}

function absmaxQuantize(weights: number[], bits = 8): AbsmaxResult {
  const levels    = (1 << (bits - 1)) - 1;   // 127 for 8-bit
  const ma        = maxAbs(weights) || 1;
  const scale     = ma / levels;

  const quantized   = new Int8Array(weights.map(w => Math.round(w / scale)));
  const dequantized = Array.from(quantized).map(q => q * scale);
  return { quantized, scale, dequantized };
}

// ──────────────────────────────────────────────────────────────────────────────
// 2. Zero-point affine quantization
// ──────────────────────────────────────────────────────────────────────────────
interface ZPResult {
  quantized:   Uint8Array;
  scale:       number;
  zeroPoint:   number;
  dequantized: number[];
}

function zeropointQuantize(weights: number[], bits = 4): ZPResult {
  const levels    = (1 << bits) - 1;           // 15 for 4-bit
  const wMin      = minOf(weights), wMax = maxOf(weights);
  const scale     = (wMax - wMin) / levels || 1;
  const zeroPoint = Math.round(-wMin / scale);

  const quantized   = new Uint8Array(weights.map(w =>
    Math.min(levels, Math.max(0, Math.round(w / scale) + zeroPoint))
  ));
  const dequantized = Array.from(quantized).map(q => (q - zeroPoint) * scale);
  return { quantized, scale, zeroPoint, dequantized };
}

// ──────────────────────────────────────────────────────────────────────────────
// 3. Block quantization
// ──────────────────────────────────────────────────────────────────────────────
interface BlockResult {
  quantized:   Int8Array;
  scales:      number[];
  dequantized: number[];
}

function blockQuantize(weights: number[], bits = 4, blockSize = 4): BlockResult {
  const levels    = (1 << (bits - 1)) - 1;
  const n         = weights.length;
  const nBlocks   = Math.ceil(n / blockSize);
  const quantized   = new Int8Array(n);
  const dequantized = new Array(n).fill(0);
  const scales: number[] = [];

  for (let b = 0; b < nBlocks; b++) {
    const start = b * blockSize, end = Math.min(start + blockSize, n);
    const block = weights.slice(start, end);
    const ma    = maxAbs(block) || 1;
    const scale = ma / levels;
    scales.push(scale);
    for (let i = start; i < end; i++) {
      const q = Math.round(weights[i] / scale);
      quantized[i]   = Math.max(-levels, Math.min(levels, q));
      dequantized[i] = quantized[i] * scale;
    }
  }
  return { quantized, scales, dequantized };
}

// ──────────────────────────────────────────────────────────────────────────────
// 4. NF4 — Normal Float 4 (QLoRA)
// ──────────────────────────────────────────────────────────────────────────────
const NF4_LEVELS: number[] = [
  -1.0, -0.6961, -0.5251, -0.3951,
  -0.2844, -0.1848, -0.0917,  0.0,
   0.0796,  0.1609,  0.2461,  0.3379,
   0.4407,  0.5626,  0.7230,  1.0,
];

interface NF4Result {
  indices:     Uint8Array;
  absmax:      number;
  dequantized: number[];
}

function nf4Quantize(weights: number[], absmax?: number): NF4Result {
  const am = absmax ?? maxAbs(weights) || 1;
  const indices = new Uint8Array(weights.map(w => {
    const norm  = w / am;
    let best = 0, bestDist = Infinity;
    for (let i = 0; i < NF4_LEVELS.length; i++) {
      const d = Math.abs(norm - NF4_LEVELS[i]);
      if (d < bestDist) { bestDist = d; best = i; }
    }
    return best;
  }));
  const dequantized = Array.from(indices).map(i => NF4_LEVELS[i] * am);
  return { indices, absmax: am, dequantized };
}

// ──────────────────────────────────────────────────────────────────────────────
// Bit inspector
// ──────────────────────────────────────────────────────────────────────────────
function toBin(n: number, bits: number): string {
  const masked = ((n | 0) + (1 << bits)) & ((1 << bits) - 1);
  return masked.toString(2).padStart(bits, "0");
}

function showBitComparison(
  original: number, quantizedInt: number, reconstructed: number, bits: number
): void {
  console.log(`    original   ${original >= 0 ? "+" : ""}${original.toFixed(6)}`);
  console.log(`    quantized  ${quantizedInt} → int${bits}: ${toBin(quantizedInt, bits)}`);
  console.log(`    recon'd    ${reconstructed >= 0 ? "+" : ""}${reconstructed.toFixed(6)}`);
  console.log(`    bit error: |${original.toFixed(6)} - ${reconstructed.toFixed(6)}| = ${Math.abs(original - reconstructed).toFixed(6)}`);
}

// ──────────────────────────────────────────────────────────────────────────────
// Main demo
// ──────────────────────────────────────────────────────────────────────────────
function main(): void {
  const HR = "=".repeat(65);
  console.log(HR);
  console.log("  Bit-Level Weight Quantization Engine — from scratch (TypeScript)");
  console.log(HR);

  const rng     = new LCG(0);
  const ROWS = 4, COLS = 8;
  const weightsMat = rng.matrix(ROWS, COLS, 0.02);
  const weights    = flatten(weightsMat);
  const n          = weights.length;
  const fp32Bytes  = memBytes(32, n);

  console.log(`\nWeight matrix shape: [${ROWS}, ${COLS}]  (${n} elements)`);
  console.log(`Original (float32) memory: ${fp32Bytes} bytes`);
  console.log("\nSample weights:");
  weightsMat.forEach(row => console.log(" ", row.map(v => v.toFixed(5)).join("  ")));

  // 1 ─ Absmax 8-bit
  console.log(`\n${"─".repeat(65)}`);
  console.log("  [1] Absmax Symmetric — 8-bit");
  console.log("─".repeat(65));
  const r8 = absmaxQuantize(weights, 8);
  console.log(`  Scale : ${r8.scale.toFixed(8)}`);
  console.log(`  MSE   : ${mse(weights, r8.dequantized).toExponential(2)}`);
  console.log(`  SNR   : ${snrDb(weights, r8.dequantized).toFixed(1)} dB`);
  console.log(`  Memory: ${memBytes(8, n)} bytes  (${(fp32Bytes / memBytes(8, n)).toFixed(1)}x smaller)`);
  console.log("\n  Bit-level inspection (first weight):");
  showBitComparison(weights[0], r8.quantized[0], r8.dequantized[0], 8);

  // 2 ─ ZeroPoint 4-bit
  console.log(`\n${"─".repeat(65)}`);
  console.log("  [2] Zero-Point Affine — 4-bit");
  console.log("─".repeat(65));
  const r4 = zeropointQuantize(weights, 4);
  console.log(`  Scale: ${r4.scale.toFixed(8)}   Zero-point: ${r4.zeroPoint}`);
  console.log(`  MSE  : ${mse(weights, r4.dequantized).toExponential(2)}`);
  console.log(`  SNR  : ${snrDb(weights, r4.dequantized).toFixed(1)} dB`);
  console.log(`  Memory: ${memBytes(4, n)} bytes  (${(fp32Bytes / memBytes(4, n)).toFixed(1)}x smaller)`);
  console.log("\n  Bit-level inspection (first weight):");
  showBitComparison(weights[0], r4.quantized[0], r4.dequantized[0], 4);

  // 3 ─ Block 4-bit
  console.log(`\n${"─".repeat(65)}`);
  console.log("  [3] Block Quantization — 4-bit, block_size=4");
  console.log("─".repeat(65));
  const rb = blockQuantize(weights, 4, 4);
  console.log(`  Per-block scales: ${rb.scales.map(s => s.toFixed(6)).join("  ")}`);
  console.log(`  MSE  : ${mse(weights, rb.dequantized).toExponential(2)}`);
  console.log(`  SNR  : ${snrDb(weights, rb.dequantized).toFixed(1)} dB`);

  // 4 ─ NF4
  console.log(`\n${"─".repeat(65)}`);
  console.log("  [4] NF4 — Normal Float 4 (QLoRA)");
  console.log("─".repeat(65));
  const rnf = nf4Quantize(weights);
  console.log(`  Absmax: ${rnf.absmax.toFixed(8)}`);
  console.log(`  MSE   : ${mse(weights, rnf.dequantized).toExponential(2)}`);
  console.log(`  SNR   : ${snrDb(weights, rnf.dequantized).toFixed(1)} dB`);
  console.log(`  Memory: ${memBytes(4, n)} bytes  (${(fp32Bytes / memBytes(4, n)).toFixed(1)}x smaller)`);

  // Summary
  console.log(`\n${HR}`);
  console.log("  Summary: precision vs. compression");
  console.log(HR);
  const rows: [string, number, number, number, number][] = [
    ["float32 (original)",  32, 0,                      Infinity,                   fp32Bytes],
    ["Absmax 8-bit",          8, mse(weights, r8.dequantized),  snrDb(weights, r8.dequantized),  memBytes(8, n)],
    ["ZeroPoint 4-bit",       4, mse(weights, r4.dequantized),  snrDb(weights, r4.dequantized),  memBytes(4, n)],
    ["Block 4-bit (b=4)",     4, mse(weights, rb.dequantized),  snrDb(weights, rb.dequantized),  memBytes(4, n)],
    ["NF4 4-bit (QLoRA)",     4, mse(weights, rnf.dequantized), snrDb(weights, rnf.dequantized), memBytes(4, n)],
  ];
  console.log(`  ${"Method".padEnd(30)} ${"Bits".padStart(4)}  ${"MSE".padStart(10)}  ${"SNR (dB)".padStart(9)}  ${"Memory".padStart(7)}`);
  console.log(`  ${"──────".padEnd(30)} ${"────".padStart(4)}  ${"───".padStart(10)}  ${"────────".padStart(9)}  ${"──────".padStart(7)}`);
  rows.forEach(([name, bits, m, s, mem]) => {
    const snrStr = isFinite(s) ? s.toFixed(1) : "∞";
    console.log(`  ${name.padEnd(30)} ${bits.toString().padStart(4)}  ${m.toExponential(2).padStart(10)}  ${snrStr.padStart(9)}  ${mem.toFixed(0).padStart(5)} B`);
  });
  console.log(HR);
}

main();
